addons/website_sale_picking/controllers/main.py

```diff
@@ -0,0 +1,23 @@
+from odoo.addons.website_sale.controllers.main import WebsiteSale
+from odoo.exceptions import UserError
+
+
+class WebsiteSaleController(WebsiteSale):
+
+    def validate_transaction_for_order(self, transaction, sale_order):
+        """
+        If this is a payment for an onsite delivery : Automatically set the order as 'sale' since the payment is not awaited before preparing the order.
+        Also sets the order shipping id to the delivery carrier's warehouse
+        """
+        super().validate_transaction_for_order(transaction, sale_order)
+
+        if transaction.acquirer_id.is_onsite_acquirer and sale_order.carrier_id.delivery_type != 'onsite':
+            raise UserError('You cannot pay onsite if the delivery is not onsite')  # This should never be triggered unless the user intentionally forges a request.
```

Comment:
ValidationError is more adapted IMO + message needs to be translated

```suggestion
            raise ValidationError(_('You cannot pay onsite if the delivery is not onsite'))  # This should never be triggered unless the user intentionally forges a request.
```